aten/src/ATen/native/Unique.cpp

```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
   return std::make_tuple(output, inverse_indices, counts);
 }
 
+// NB: Unique implementation using sort
+//
+// The whole algo is taken from NumPy at numpy/lib/arraysetops.py
+// which firstly do sort on the input sequence and then convert
+// it to consecutive unique.
+//
+// Also improvement has been made upon the NumPy version: parallel
+// `inverse_indices` and `counts` computation in a fused loop,
+// which made this part almost a free launch.
+//
+// This kernel also implements a `equal_nan` flag which has same
+// function as NumPy's unique. Currently this is always disabled.
```

Comment:
What's the behavior of CUDA implementation? Are CPU and CUDA implementation aligned? There is an open issue https://github.com/pytorch/pytorch/issues/94106 about the issue with nan handling. Can this PR address that problem?

Reply:
right now `torch.unique` treates NaN as different values (both CPU and CUDA), this behavior is different from numpy.unique which has a `equal_nan` argument. This patch won't change the behavior of `torch.unique`, but I also implemented `equal_nan` in the kernel so if someone decides that this feature shall be enabled on torch, it can be done very easily.